packages/gatsby/src/commands/serve.js

```diff
@@ -1,13 +1,39 @@
 /* @flow weak */
 const path = require(`path`)
 const openurl = require(`better-opn`)
+const fs = require(`fs-extra`)
 const signalExit = require(`signal-exit`)
 const compression = require(`compression`)
 const express = require(`express`)
-const historyFallback = require(`express-history-api-fallback`)
 const getConfigFile = require(`../bootstrap/get-config-file`)
 const preferDefault = require(`../bootstrap/prefer-default`)
 const chalk = require(`chalk`)
+const { match: reachMatch } = require(`@reach/router/lib/utils`)
+
+const getPages = directory =>
+  fs
+    .readFile(path.join(directory, `.cache`, `pages.json`))
+    .then(contents => JSON.parse(contents))
+    .catch(() => [])
+
+const clientOnlyPathsRouter = (pages, options) => {
+  const clientOnlyRoutes = pages
+    .filter(page => page.matchPath)
+    .map(page => page.matchPath)
+  return (req, res, next) => {
+    const { url } = req
+    if (req.accepts(`html`)) {
+      if (clientOnlyRoutes.some(route => reachMatch(route, url) !== null)) {
+        return res.sendFile(`index.html`, options, err => {
```

Comment:
I think we need to serve potentially other files than `index.html`

if non-index page has client only paths:

```js
createPage({
  path: '/nested',
  matchPath: '/nested/*',
  ...restOfArgs
})
```
we would want to serve `nested/index.html` for `/nested/*` paths

Reply:
```
onst clientOnlyPathsRouter = (pages, options) => {
  console.log(pages)
  const clientOnlyRoutes = pages
    .filter(page => page.matchPath)
    .map(page => ({
      path: page.path,
      matchPath: page.matchPath,
    }))
  return (req, res, next) => {
    const { url } = req
    if (req.accepts(`html`)) {
      const route = clientOnlyRoutes.find(
        route => reachMatch(route.matchPath, url) !== null
      )
      if (route) {
        return res.sendFile(`.${route.path}index.html`, options, err => {
          if (err) {
            next()
          }
        })
      }
    }
    return next()
  }
}
```

something like this might work, path is `/` for default routes and `/nested/` for the nested once. I tried it with examples/simple-auth